### PR TITLE
Single Recipe Deletion for chem dispensers

### PR DESCRIPTION
**Pull request by Liver-1**
## About The Pull Request

Adds a button to chem dispensers that allows the user to delete a single, specific recipe.
Designed with multiple checks to avoid accidental deletion, including the user having to type out the recipe in question, and a Yes/No prompt to make doubly sure. No chance of accidentally griefing yourself by clicking the wrong recipe in a rush.

## Why It's Good For The Game

It allows more granular deletion of chem dispenser macros than the current fairly limited overwrite and delete-all functionality.

## Testing Photographs and Procedure
<details>
<summary>Screenshots&Videos</summary>

- Here's the button (thanks to Babel Finger for the helpful circle and arrows)
image

- Checking the recipe buttons still work (they do!)
image

- Tested Making and using multiple recipes, in a make-test-make-test-make-make-test-test pattern. all seems to be working nicely
image

- Tested overwriting, that still works.
- tested specific deletion:
image
image
image
image
Presto.

- Tested the clear recipes button:
image
image
That works too.

</details>

## Changelog
:cl:colonelorion
add: Added Delete button to the chem dispenser
/:cl:

### CODE REVIEW

**Comment by Autisem** [comment]
I think just a yes/no prompt suffices

**Comment by Harranhall** [comment]
It'll take a bit of work, I know how to call code in the .dm file from the .js file, but I'm not sure how to go the other way. Is this good enough to use for now, and I'll implement a simple "click delete > click the recipe > press Yes" process in a further PR?

**Comment by Harranhall** [comment]
Or I could remove the Yes/No and we let the typing of the recipe be the confirmation?

**Comment by EvilDragonfiend** [comment]
Once you click `Delete`, make all recipe buttons to red/orange. That'll be intuitive than typing its name

**Comment by Harranhall** [comment]
I will try my best to figure out how to actually code that, haha. But can we put this in place for now? I don't think it's a bad change as it is, I or others can tweak it moving forward if needs be.

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm [hunk 1]
```diff
@@ -305,6 +305,18 @@
 		if("record_recipe")
 			recording_recipe = list()
 			. = TRUE
+		if("delete_recipe")
+			var/deletion_target = stripped_input(usr,"Name","please type the recipe you wish to delete", "Recipe", MAX_NAME_LEN)
```
Comment (ike709) [inline code/modules/reagents/chemistry/machinery/chem_dispenser.dm]: Just use `saved_recipes` as a list `input()`...

**Review by ike709 — CHANGES_REQUESTED** [review]
Replace the text input with a list input

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm (hunk 1 above)
Comment (Liver-1) [reply code/modules/reagents/chemistry/machinery/chem_dispenser.dm]: Working on it.

Comment (Liver-1) [reply code/modules/reagents/chemistry/machinery/chem_dispenser.dm]: How's that?
image

**Review by itsmeow — CHANGES_REQUESTED** [review]
As simple as this is, I would prefer a slightly more involved solution using TGUI states.

When you click Delete, it should enable a state value that turns the recipe buttons red, and then changes their actions to one that deletes the recipe, rather than opening a list, since the list is already on the UI it doesn't make sense to pop up another just for this.

**Comment by Harranhall** [comment]
> As simple as this is, I would prefer a slightly more involved solution using TGUI states.
> 
> When you click Delete, it should enable a state value that turns the recipe buttons red, and then changes their actions to one that deletes the recipe, rather than opening a list, since the list is already on the UI it doesn't make sense to pop up another just for this.

I agree tbh, having all the recipe buttons turn red and just clicking them would be great.

**Review by ike709 — APPROVED** [review]

**Comment by Harranhall** [comment]
Seeing as how the current version is approved, and I just want to be able to use this function in-game, I'll complete this PR with Ike's approval and look into tgui_input_list etc later.

**Comment by itsmeow** [comment]
> happy for this to merge as is, and I'll look into tgui_input_list later?

I would still prefer if it was done like this in the first place rather than another PR to just replace this PR entirely.

**Pull request closed without merging**